File: src/ChatApp.Infrastructure/Services/AdminService.cs
using AutoMapper;
using ChatApp.Application.Interfaces;
using ChatApp.Contracts.Dtos;
using ChatApp.Domain.Entities;
using ChatApp.Domain.Enums;
using ChatApp.Domain.Exceptions;
using ChatApp.Domain.ValueObjects;
using ChatApp.Infrastructure.Persistence;
using Microsoft.EntityFrameworkCore;

namespace ChatApp.Infrastructure.Services;

public class AdminService : IAdminService
{
    private readonly ChatAppDbContext _db;
    private readonly IMapper _mapper;
    private readonly IFileStorage _storage;

    public AdminService(ChatAppDbContext db, IMapper mapper, IFileStorage storage)
    {
        _db = db;
        _mapper = mapper;
        _storage = storage;
    }

    public async Task<DashboardStatsDto> GetDashboardStatsAsync(CancellationToken ct = default)
    {
        var totalUsers = await _db.Users.AsNoTracking().CountAsync(u => u.DeletedAt == null, ct);
        var onlineUsers = await _db.Users.AsNoTracking().CountAsync(u => u.IsOnline && u.DeletedAt == null, ct);
        var totalConversations = await _db.Conversations.AsNoTracking().CountAsync(c => c.DeletedAt == null, ct);
        var totalMessages = await _db.Messages.AsNoTracking().CountAsync(m => m.DeletedAt == null, ct);
        var totalAttachments = await _db.MessageAttachments.AsNoTracking().CountAsync(ct);
        var totalAdmins = await _db.Users.AsNoTracking().CountAsync(u => u.Role == UserRole.Admin && u.DeletedAt == null, ct);
        var disabledUsers = await _db.Users.AsNoTracking().CountAsync(u => u.Status == UserStatus.Disabled && u.DeletedAt == null, ct);
        var activeUploads = await _db.FileUploads.AsNoTracking().CountAsync(f => f.Status == UploadStatus.Uploading || f.Status == UploadStatus.Pending, ct);

        long totalSize = 0;
        var sizes = await _db.FileUploads.AsNoTracking().Where(f => f.Status == UploadStatus.Completed).Select(f => f.Size).ToListAsync(ct);
        totalSize = sizes.Sum();

        return new DashboardStatsDto(
            totalUsers, onlineUsers, totalConversations, totalMessages,
            totalAttachments, totalSize, totalAdmins, disabledUsers, activeUploads);
    }

    public async Task<IReadOnlyList<UserDto>> ListUsersAsync(string? search, string? phone, int page, int pageSize, CancellationToken ct = default)
    {
        page = Math.Max(1, page);
        pageSize = Math.Clamp(pageSize, 1, 200);
        var q = _db.Users.AsNoTracking().Where(u => u.DeletedAt == null);
        if (!string.IsNullOrWhiteSpace(search))
        {
            var s = search.Trim();
            q = q.Where(u => u.FirstName.Contains(s) || u.LastName.Contains(s) || u.FullName.Contains(s));
        }
        if (!string.IsNullOrWhiteSpace(phone))
        {
            var p = phone.Trim();
            // Try normalize
            if (PhoneNumber.TryParse(p, out var pn))
                q = q.Where(u => u.NormalizedPhoneNumber == pn.E164);
            else
                q = q.Where(u => u.PhoneNumber.Contains(p));
        }
        var users = await q.OrderByDescending(u => u.CreatedAt).Skip((page - 1) * pageSize).Take(pageSize).ToListAsync(ct);
        return _mapper.Map<List<UserDto>>(users);
    }

    public async Task<UserDto?> GetUserAsync(Guid id, CancellationToken ct = default)
    {
        var user = await _db.Users.AsNoTracking().FirstOrDefaultAsync(u => u.Id == id, ct);
        return user is null ? null : _mapper.Map<UserDto>(user);
    }

    public async Task<UserDto> ChangeUserRoleAsync(Guid adminId, Guid userId, string role, string? ipAddress, CancellationToken ct = default)
    {
        if (!Enum.TryParse<UserRole>(role, true, out var r)) throw new DomainException($"نقش نامعتبر: {role}");
        var user = await _db.Users.FirstOrDefaultAsync(u => u.Id == userId, ct) ?? throw new EntityNotFoundException("User", userId);
        var oldRole = user.Role;
        user.Role = r;
        user.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync(ct);

        _db.AuditLogs.Add(new AuditLog
        {
            AdminId = adminId,
            Action = AuditAction.ChangeRole,
            TargetUserId = userId,
            Details = $"Role changed from {oldRole} to {r}",
            IpAddress = ipAddress,
            CreatedAt = DateTime.UtcNow
        });
        await _db.SaveChangesAsync(ct);
        return _mapper.Map<UserDto>(user);
    }

    public async Task<UserDto> ChangeUserStatusAsync(Guid adminId, Guid userId, string status, string? ipAddress, CancellationToken ct = default)
    {
        if (!Enum.TryParse<UserStatus>(status, true, out var s)) throw new DomainException($"وضعیت نامعتبر: {status}");
        var user = await _db.Users.FirstOrDefaultAsync(u => u.Id == userId, ct) ?? throw new EntityNotFoundException("User", userId);
        var oldStatus = user.Status;
        user.Status = s;
        user.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync(ct);

        _db.AuditLogs.Add(new AuditLog
        {
            AdminId = adminId,
            Action = s == UserStatus.Disabled ? AuditAction.DisableUser : (s == UserStatus.Active ? AuditAction.EnableUser : AuditAction.DeleteUser),
            TargetUserId = userId,
            Details = $"Status changed from {oldStatus} to {s}",
            IpAddress = ipAddress,
            CreatedAt = DateTime.UtcNow
        });
        await _db.SaveChangesAsync(ct);
        return _mapper.Map<UserDto>(user);
    }

    public async Task<IReadOnlyList<ConversationDto>> ListUserConversationsAsync(Guid adminId, Guid userId, string? ipAddress, int page, int pageSize, CancellationToken ct = default)
    {
        page = Math.Max(1, page);
        pageSize = Math.Clamp(pageSize, 1, 100);

        var conversations = await (
            from c in _db.Conversations.AsNoTracking()
            where c.DeletedAt == null && c.Participants.Any(p => p.UserId == userId)
            orderby c.LastMessageAt ?? c.CreatedAt descending
            select c
        ).Skip((page - 1) * pageSize).Take(pageSize).ToListAsync(ct);

        var result = new List<ConversationDto>();
        foreach (var c in conversations)
        {
            var dto = _mapper.Map<ConversationDto>(c);
            var others = await (
                from p in _db.ConversationParticipants.AsNoTracking()
                join u in _db.Users.AsNoTracking() on p.UserId equals u.Id
                where p.ConversationId == c.Id && p.UserId != userId
                select u
            ).FirstOrDefaultAsync(ct);
            dto.OtherParticipant = others is not null ? _mapper.Map<UserSummaryDto>(others) : null;
            dto.UnreadCount = 0;
            result.Add(dto);
        }

        _db.AuditLogs.Add(new AuditLog
        {
            AdminId = adminId,
            Action = AuditAction.ViewUser,
            TargetUserId = userId,
            Details = "Admin viewed user conversations",
            IpAddress = ipAddress,
            CreatedAt = DateTime.UtcNow
        });
        await _db.SaveChangesAsync(ct);
        return result;
    }

    public async Task<IReadOnlyList<MessageDto>> ListConversationMessagesAsync(Guid adminId, Guid conversationId, string? ipAddress, int page, int pageSize, CancellationToken ct = default)
    {
        page = Math.Max(1, page);
        pageSize = Math.Clamp(pageSize, 1, 200);

        var conv = await _db.Conversations.AsNoTracking().FirstOrDefaultAsync(c => c.Id == conversationId, ct);
        if (conv is null) return Array.Empty<MessageDto>();

        var msgs = await _db.Messages.AsNoTracking()
            .Where(m => m.ConversationId == conversationId && m.DeletedAt == null)
            .OrderByDescending(m => m.CreatedAt)
            .Skip((page - 1) * pageSize).Take(pageSize)
            .ToListAsync(ct);

        var senders = await _db.Users.AsNoTracking().Where(u => msgs.Select(m => m.SenderId).Distinct().Contains(u.Id)).ToDictionaryAsync(u => u.Id, ct);
        var attachments = await _db.MessageAttachments.AsNoTracking().Where(a => msgs.Select(m => m.Id).Contains(a.MessageId)).ToListAsync(ct);

        var result = msgs.Select(m => new MessageDto(
            Id: m.Id,
            ConversationId: m.ConversationId,
            SenderId: m.SenderId,
            SenderName: senders.TryGetValue(m.SenderId, out var s) ? s.FullName : "",
            SenderAvatarUrl: senders.TryGetValue(m.SenderId, out var s2) ? s2.AvatarUrl : null,
            Content: m.Content,
            Type: m.Type.ToString(),
            Status: m.Status.ToString(),
            ReplyToMessageId: m.ReplyToMessageId,
            ReplyToPreview: null,
            CreatedAt: m.CreatedAt,
            UpdatedAt: m.UpdatedAt,
            DeletedAt: m.DeletedAt,
            IsEdited: m.IsEdited,
            DeliveredAt: m.DeliveredAt,
            ReadAt: m.ReadAt,
            Attachments: attachments.Where(a => a.MessageId == m.Id).Select(a => new AttachmentDto(
                Id: a.Id, OriginalFileName: a.OriginalFileName, Size: a.Size,
                ContentType: a.ContentType, Type: a.Type.ToString(),
                ThumbnailUrl: a.ThumbnailPath, DownloadUrl: $"/api/files/{a.Id}")).ToList()
        )).ToList();

        _db.AuditLogs.Add(new AuditLog
        {
            AdminId = adminId,
            Action = AuditAction.ViewConversation,
            TargetConversationId = conversationId,
            Details = $"Admin viewed conversation messages ({result.Count})",
            IpAddress = ipAddress,
            CreatedAt = DateTime.UtcNow
        });
        await _db.SaveChangesAsync(ct);
        return result;
    }

    public async Task<IReadOnlyList<AuditLogDto>> ListAuditLogsAsync(DateTime? from, DateTime? to, int page, int pageSize, CancellationToken ct = default)
    {
        page = Math.Max(1, page);
        pageSize = Math.Clamp(pageSize, 1, 200);
        var q = _db.AuditLogs.AsNoTracking();
        if (from.HasValue) q = q.Where(a => a.CreatedAt >= from.Value);
        if (to.HasValue) q = q.Where(a => a.CreatedAt <= to.Value);

        var logs = await q.OrderByDescending(a => a.CreatedAt).Skip((page - 1) * pageSize).Take(pageSize).ToListAsync(ct);
        var adminIds = logs.Select(l => l.AdminId).Distinct().ToList();
        var admins = await _db.Users.AsNoTracking().Where(u => adminIds.Contains(u.Id)).ToDictionaryAsync(u => u.Id, ct);
        var targetIds = logs.Where(l => l.TargetUserId.HasValue).Select(l => l.TargetUserId!.Value).Distinct().ToList();
        var targets = targetIds.Count > 0
            ? await _db.Users.AsNoTracking().Where(u => targetIds.Contains(u.Id)).ToDictionaryAsync(u => u.Id, ct)
            : new Dictionary<Guid, User>();

        return logs.Select(l => new AuditLogDto(
            Id: l.Id,
            AdminId: l.AdminId,
            AdminName: admins.TryGetValue(l.AdminId, out var a) ? a.FullName : "",
            Action: l.Action.ToString(),
            TargetUserId: l.TargetUserId,
            TargetUserName: l.TargetUserId.HasValue && targets.TryGetValue(l.TargetUserId.Value, out var t) ? t.FullName : null,
            TargetConversationId: l.TargetConversationId,
            TargetMessageId: l.TargetMessageId,
            TargetAttachmentId: l.TargetAttachmentId,
            Details: l.Details,
            IpAddress: l.IpAddress,
            CreatedAt: l.CreatedAt
        )).ToList();
    }

    public async Task UpdateAdminPhoneAsync(Guid adminId, string newPhone, string? ipAddress, CancellationToken ct = default)
    {
        var phone = PhoneNumber.Create(newPhone);
        var setting = await _db.AppSettings.FirstOrDefaultAsync(s => s.Key == "INITIAL_ADMIN_PHONE", ct);
        if (setting is null)
        {
            setting = new AppSetting
            {
                Key = "INITIAL_ADMIN_PHONE",
                Value = phone.E164,
                Description = "Phone number of the initial admin user (granted Admin role on registration).",
                UpdatedAt = DateTime.UtcNow
            };
            _db.AppSettings.Add(setting);
        }
        else
        {
            setting.Value = phone.E164;
            setting.UpdatedAt = DateTime.UtcNow;
        }

        _db.AuditLogs.Add(new AuditLog
        {
            AdminId = adminId,
            Action = AuditAction.ChangeAdmin,
            Details = $"Admin phone changed to {phone.E164}",
            IpAddress = ipAddress,
            CreatedAt = DateTime.UtcNow
        });
        await _db.SaveChangesAsync(ct);
    }

    public async Task<string> GetAdminPhoneAsync(CancellationToken ct = default)
    {
        var setting = await _db.AppSettings.AsNoTracking().FirstOrDefaultAsync(s => s.Key == "INITIAL_ADMIN_PHONE", ct);
        return setting?.Value ?? string.Empty;
    }

    public async Task<IReadOnlyList<MessageDto>> SearchMessagesAsync(Guid adminId, string query, int page, int pageSize, CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(query)) return Array.Empty<MessageDto>();
        page = Math.Max(1, page);
        pageSize = Math.Clamp(pageSize, 1, 200);
        var q = query.Trim();

        var msgs = await _db.Messages.AsNoTracking()
            .Where(m => m.DeletedAt == null && m.Content.Contains(q))
            .OrderByDescending(m => m.CreatedAt)
            .Skip((page - 1) * pageSize).Take(pageSize)
            .ToListAsync(ct);

        var senders = await _db.Users.AsNoTracking().Where(u => msgs.Select(m => m.SenderId).Distinct().Contains(u.Id)).ToDictionaryAsync(u => u.Id, ct);

        var result = msgs.Select(m => new MessageDto(
            Id: m.Id,
            ConversationId: m.ConversationId,
            SenderId: m.SenderId,
            SenderName: senders.TryGetValue(m.SenderId, out var s) ? s.FullName : "",
            SenderAvatarUrl: senders.TryGetValue(m.SenderId, out var s2) ? s2.AvatarUrl : null,
            Content: m.Content,
            Type: m.Type.ToString(),
            Status: m.Status.ToString(),
            ReplyToMessageId: m.ReplyToMessageId,
            ReplyToPreview: null,
            CreatedAt: m.CreatedAt,
            UpdatedAt: m.UpdatedAt,
            DeletedAt: m.DeletedAt,
            IsEdited: m.IsEdited,
            DeliveredAt: m.DeliveredAt,
            ReadAt: m.ReadAt,
            Attachments: new List<AttachmentDto>()
        )).ToList();

        _db.AuditLogs.Add(new AuditLog
        {
            AdminId = adminId,
            Action = AuditAction.ViewMessage,
            Details = $"Admin searched messages for '{query}'",
            IpAddress = null,
            CreatedAt = DateTime.UtcNow
        });
        await _db.SaveChangesAsync(ct);
        return result;
    }
}

File: backend/src/admin/services/admin.service.ts
import { Injectable, NotFoundException, BadRequestException, ForbiddenException, Logger } from '@nestjs/common';
import { PrismaService } from '../../prisma/prisma.service';
import { PhoneUtil } from '../../common/utils/phone.util';
import {
  AdminListUsersQueryDto,
  UpdateUserRoleDto,
  UpdateUserStatusDto,
  UpdateAdminPhoneDto,
  AdminListConversationsQueryDto,
  AdminListMessagesQueryDto,
  AdminSearchMessagesQueryDto,
  AdminAuditLogsQueryDto,
} from '../dto/admin.dto';

@Injectable()
export class AdminService {
  private readonly logger = new Logger('AdminService');

  constructor(private readonly prisma: PrismaService) {}

  async getDashboardStats() {
    const [
      totalUsers,
      onlineUsers,
      totalConversations,
      totalMessages,
      totalAttachments,
      totalAdmins,
      disabledUsers,
      activeUploads,
      attachmentSizes,
    ] = await Promise.all([
      this.prisma.user.count({ where: { deletedAt: null } }),
      this.prisma.user.count({ where: { isOnline: true, deletedAt: null } }),
      this.prisma.conversation.count({ where: { deletedAt: null } }),
      this.prisma.message.count({ where: { deletedAt: null } }),
      this.prisma.attachment.count(),
      this.prisma.user.count({ where: { role: 'ADMIN', deletedAt: null } }),
      this.prisma.user.count({ where: { status: 'DISABLED', deletedAt: null } }),
      this.prisma.attachment.count({ where: { uploadStatus: { in: ['UPLOADING', 'PENDING'] } } }),
      this.prisma.attachment.aggregate({ where: { uploadStatus: 'COMPLETED' }, _sum: { size: true } }),
    ]);

    return {
      totalUsers,
      onlineUsers,
      totalConversations,
      totalMessages,
      totalAttachments,
      totalAttachmentSizeBytes: (attachmentSizes._sum.size || BigInt(0)).toString(),
      totalAdmins,
      disabledUsers,
      activeUploads,
    };
  }

  async listUsers(q: AdminListUsersQueryDto) {
    const where: any = { deletedAt: null };
    if (q.search) {
      const s = q.search.trim();
      where.OR = [
        { firstName: { contains: s } },
        { lastName: { contains: s } },
        { fullName: { contains: s } },
      ];
    }
    if (q.phone) {
      const normalized = PhoneUtil.tryNormalize(q.phone);
      if (normalized) where.normalizedPhone = normalized;
      else where.phoneNumber = { contains: q.phone };
    }

    const [items, total] = await Promise.all([
      this.prisma.user.findMany({
        where,
        orderBy: { createdAt: 'desc' },
        skip: (q.page - 1) * q.pageSize,
        take: q.pageSize,
        select: {
          id: true, firstName: true, lastName: true, fullName: true,
          phoneNumber: true, avatarUrl: true, role: true, status: true,
          isOnline: true, lastSeen: true, createdAt: true,
        },
      }),
      this.prisma.user.count({ where }),
    ]);

    return { items, total, page: q.page, pageSize: q.pageSize };
  }

  async getUser(id: string) {
    const user = await this.prisma.user.findUnique({
      where: { id },
      select: {
        id: true, firstName: true, lastName: true, fullName: true,
        phoneNumber: true, avatarUrl: true, bio: true, role: true, status: true,
        isOnline: true, lastSeen: true, createdAt: true, updatedAt: true,
      },
    });
    if (!user) throw new NotFoundException('کاربر یافت نشد.');
    return user;
  }

  async changeUserRole(adminId: string, targetUserId: string, dto: UpdateUserRoleDto, ip?: string) {
    if (!['USER', 'ADMIN'].includes(dto.role))
      throw new BadRequestException('نقش نامعتبر است.');
    if (adminId === targetUserId)
      throw new ForbiddenException('نمی‌توانید نقش خودتان را تغییر دهید.');

    const user = await this.prisma.user.findUnique({ where: { id: targetUserId } });
    if (!user) throw new NotFoundException('کاربر یافت نشد.');

    const oldRole = user.role;
    const updated = await this.prisma.user.update({
      where: { id: targetUserId },
      data: { role: dto.role },
    });

    await this.prisma.auditLog.create({
      data: {
        adminId,
        action: 'CHANGE_ROLE',
        targetUserId,
        details: `Role changed from ${oldRole} to ${dto.role}`,
        ipAddress: ip,
      },
    });

    return this.toUserDto(updated);
  }

  async changeUserStatus(adminId: string, targetUserId: string, dto: UpdateUserStatusDto, ip?: string) {
    if (!['ACTIVE', 'DISABLED', 'DELETED'].includes(dto.status))
      throw new BadRequestException('وضعیت نامعتبر است.');
    if (adminId === targetUserId)
      throw new ForbiddenException('نمی‌توانید وضعیت خودتان را تغییر دهید.');

    const user = await this.prisma.user.findUnique({ where: { id: targetUserId } });
    if (!user) throw new NotFoundException('کاربر یافت نشد.');

    const oldStatus = user.status;
    const updated = await this.prisma.user.update({
      where: { id: targetUserId },
      data: { status: dto.status, deletedAt: dto.status === 'DELETED' ? new Date() : user.deletedAt },
    });

    const actionMap: Record<string, string> = {
      ACTIVE: 'ENABLE_USER',
      DISABLED: 'DISABLE_USER',
      DELETED: 'DELETE_USER',
    };

    await this.prisma.auditLog.create({
      data: {
        adminId,
        action: actionMap[dto.status] || 'UPDATE_SETTINGS',
        targetUserId,
        details: `Status changed from ${oldStatus} to ${dto.status}`,
        ipAddress: ip,
      },
    });

    return this.toUserDto(updated);
  }

  async listUserConversations(adminId: string, userId: string, q: AdminListConversationsQueryDto, ip?: string) {
    const [items, total] = await Promise.all([
      this.prisma.conversation.findMany({
        where: { deletedAt: null, participants: { some: { userId } } },
        include: {
          participants: { include: { user: { select: { id: true, fullName: true, phoneNumber: true, avatarUrl: true, isOnline: true, lastSeen: true } } } },
        },
        orderBy: [{ lastMessageAt: 'desc' }, { createdAt: 'desc' }],
        skip: (q.page - 1) * q.pageSize,
        take: q.pageSize,
      }),
      this.prisma.conversation.count({ where: { deletedAt: null, participants: { some: { userId } } } }),
    ]);

    await this.prisma.auditLog.create({
      data: {
        adminId,
        action: 'VIEW_USER',
        targetUserId: userId,
        details: 'Admin viewed user conversations',
        ipAddress: ip,
      },
    });

    return {
      items: items.map((c) => ({
        id: c.id,
        isGroup: c.isGroup,
        title: c.title,
        lastMessageAt: c.lastMessageAt,
        lastMessagePreview: c.lastMessagePreview,
        createdAt: c.createdAt,
        otherParticipant: c.participants.find((p) => p.userId !== userId)?.user || null,
      })),
      total,
      page: q.page,
      pageSize: q.pageSize,
    };
  }

  async listConversationMessages(adminId: string, conversationId: string, q: AdminListMessagesQueryDto, ip?: string) {
    const [items, total] = await Promise.all([
      this.prisma.message.findMany({
        where: { conversationId, deletedAt: null },
        include: {
          sender: { select: { id: true, fullName: true, avatarUrl: true } },
          attachments: true,
        },
        orderBy: { createdAt: 'desc' },
        skip: (q.page - 1) * q.pageSize,
        take: q.pageSize,
      }),
      this.prisma.message.count({ where: { conversationId, deletedAt: null } }),
    ]);

    await this.prisma.auditLog.create({
      data: {
        adminId,
        action: 'VIEW_CONVERSATION',
        targetConversationId: conversationId,
        details: `Admin viewed conversation messages (${items.length})`,
        ipAddress: ip,
      },
    });

    return {
      items: items.map((m) => ({
        id: m.id,
        conversationId: m.conversationId,
        senderId: m.senderId,
        senderName: m.sender?.fullName || '',
        senderAvatarUrl: m.sender?.avatarUrl || null,
        content: m.content,
        type: m.type,
        status: m.status,
        createdAt: m.createdAt,
        deletedAt: m.deletedAt,
        attachments: m.attachments.map((a) => ({
          id: a.id,
          originalFileName: a.originalFileName,
          size: a.size.toString(),
          mimeType: a.mimeType,
          type: a.type,
          downloadUrl: `/api/files/${a.id}`,
        })),
      })),
      total,
      page: q.page,
      pageSize: q.pageSize,
    };
  }

  async searchMessages(adminId: string, q: AdminSearchMessagesQueryDto, ip?: string) {
    const s = q.q.trim();
    if (!s) return { items: [], total: 0 };

    const [items, total] = await Promise.all([
      this.prisma.message.findMany({
        where: { deletedAt: null, content: { contains: s } },
        include: { sender: { select: { id: true, fullName: true } }, conversation: true },
        orderBy: { createdAt: 'desc' },
        skip: (q.page - 1) * q.pageSize,
        take: q.pageSize,
      }),
      this.prisma.message.count({ where: { deletedAt: null, content: { contains: s } } }),
    ]);

    await this.prisma.auditLog.create({
      data: {
        adminId,
        action: 'VIEW_MESSAGE',
        details: `Admin searched messages for "${s}"`,
        ipAddress: ip,
      },
    });

    return {
      items: items.map((m) => ({
        id: m.id,
        conversationId: m.conversationId,
        senderId: m.senderId,
        senderName: m.sender?.fullName || '',
        content: m.content,
        type: m.type,
        status: m.status,
        createdAt: m.createdAt,
      })),
      total,
      page: q.page,
      pageSize: q.pageSize,
    };
  }

  async listAuditLogs(q: AdminAuditLogsQueryDto) {
    const where: any = {};
    if (q.from) where.createdAt = { gte: new Date(q.from) };
    if (q.to) where.createdAt = where.createdAt
      ? { ...where.createdAt, lte: new Date(q.to) }
      : { lte: new Date(q.to) };

    const [items, total] = await Promise.all([
      this.prisma.auditLog.findMany({
        where,
        include: { admin: { select: { fullName: true } } },
        orderBy: { createdAt: 'desc' },
        skip: (q.page - 1) * q.pageSize,
        take: q.pageSize,
      }),
      this.prisma.auditLog.count({ where }),
    ]);

    return {
      items: items.map((l) => ({
        id: l.id,
        adminId: l.adminId,
        adminName: l.admin?.fullName || '',
        action: l.action,
        targetUserId: l.targetUserId,
        targetConversationId: l.targetConversationId,
        targetMessageId: l.targetMessageId,
        targetAttachmentId: l.targetAttachmentId,
        details: l.details,
        ipAddress: l.ipAddress,
        createdAt: l.createdAt,
      })),
      total,
      page: q.page,
      pageSize: q.pageSize,
    };
  }

  async getAdminPhone() {
    const s = await this.prisma.appSetting.findUnique({ where: { key: 'INITIAL_ADMIN_PHONE' } });
    return { phoneNumber: s?.value || '' };
  }

  async updateAdminPhone(adminId: string, dto: UpdateAdminPhoneDto, ip?: string) {
    const phone = PhoneUtil.normalize(dto.phoneNumber);
    await this.prisma.appSetting.upsert({
      where: { key: 'INITIAL_ADMIN_PHONE' },
      update: { value: phone },
      create: {
        key: 'INITIAL_ADMIN_PHONE',
        value: phone,
        description: 'Phone number of the initial admin user.',
      },
    });

    await this.prisma.auditLog.create({
      data: {
        adminId,
        action: 'CHANGE_ADMIN',
        details: `Admin phone changed to ${phone}`,
        ipAddress: ip,
      },
    });

    return { success: true, phoneNumber: phone };
  }

  private toUserDto(u: any) {
    return {
      id: u.id,
      firstName: u.firstName,
      lastName: u.lastName,
      fullName: u.fullName,
      phoneNumber: u.phoneNumber,
      avatarUrl: u.avatarUrl,
      role: u.role,
      status: u.status,
      isOnline: u.isOnline,
      lastSeen: u.lastSeen,
      createdAt: u.createdAt,
      updatedAt: u.updatedAt,
    };
  }
}
